ms_teams_webclient.py

```diff
@@ -0,0 +1,153 @@
+        """Post response to callback url
+
+        Send a reply to URL indicated in serviceUrl from
+        Bot Framework's request object.
+
+        @param response: activity object
+        """
+        r = requests.post(
+            response.post_url,
+            data=json.dumps(response.payload),
+            headers=self.__get_default_headers()
+        )
+
+        r.raise_for_status()
+
+    def add_reaction(self, message, reaction):
+        emoji = reactions[reaction] if reactions[reaction] else ''
+        message.body = f'> **@{message.frm.nick}**: {message.body}\n\n{emoji}'
+        reply = self.build_reply(message)
+        self.send_reply(reply)
+
+    def get_token(self):
+        now = datetime.datetime.now()
+        if not self.__token or self.__token.expired_at <= now:
+            self.__token = self.__auth()
+        return self.__token.access_token
+    
+    def build_reply(self, message):
+        req = message.extras['conversation']
+        payload = {
+            'type': 'message',
+            'conversation': req.conversation,
+            'from': message.to.subject,
+            'recipient': message.frm.subject,
+            'replyToId': req.conversation_id,
+            'text': message.body,
+            'textFormat': 'markdown'
+        }
+        return activity(req.reply_url, payload)
+
+    def __get_member_by_email(self, identifier):
+        service_url = identifier.extras['service_url']
+        team_id = identifier.extras['team_id']
+        email = identifier.email
+
+        response = requests.get(
+            f'{service_url}/v3/conversations/{team_id}/members/{email}',
+            headers=self.__get_default_headers()
+        )
+        if response.status_code == 404:
+            raise MemberNotFound(f"member not found using {identifier.email} email")
+        return response.json()
+
+    def __auth(self):
+        form = {
+            'grant_type': 'client_credentials',
+            'scope': 'https://api.botframework.com/.default',
+            'client_id': self.__app_id,
+            'client_secret': self.__app_password,
+        }
+
+        r = requests.post(
+            'https://login.microsoftonline.com/botframework.com/oauth2/v2.0/token',
+            data=form
+        ).json()
+
+        expires_in = r['expires_in']
+        expired_at = from_now(expires_in)
+        token = authtoken(r['access_token'], expired_at)
+
+        return token
+
+    def __send_direct_message(self, text, extras):
+        body = {
+            "type": "message",
+            "text": text
+        }
+        response = requests.post(
+            f'{extras["service_url"]}/v3/conversations/{extras["conversation"]["id"]}/activities/',
+            json.dumps(body),
+            headers=self.__get_default_headers()
+        )
+
+        return response.json()
+
+    def __create_conversation(self, member_id, extras):
+        body = {
+            "bot": {
+                "id": f"28:{self.__app_id}"
```

Comment:
what is 28?

Reply:
Good point! This prefix is needed for Azure Bot implementations. We added a new constant for making it easier to understand.